drivers: led: lp50xx: add support for single color LEDs

ivanwagner opened the pull request:
Max LED quantities are based on the assumption that RGB LEDs are being used. This patch adds support for arbitrary LEDs (e.g. single classical LEDs) and limit check is being performed by reading the defined color-mapping field within the device tree definition.

ivanwagner commented:
> > Max LED quantities are based on the assumption that RGB LEDs are being used. This patch adds support for arbitrary LEDs (e.g. single classical LEDs) and limit check is being performed by reading the defined color-mapping field within the device tree definition.
> 
> Hi @ivanwagner,
> 
> I think that the manufacturer (Texas Instruments) also made the same assumption that RGB LEDs would be used with these controllers :) For example the datasheet of the LP503x controllers calls them "RGB LED drivers". While it provides 30 or 36 output color registers, it only provides 10 or 12 brightness registers (one per LED).
> 
> So maybe this PR adds a use case to the driver that it should not support.

Hi @simonguinot,

first of all thank you for your reply. Although it is true that the brightness registers are only 8 on the lp5024 device, the data-sheet also states the following at page 13:

> Keeping FFh (default value) in the LED0_BRIGHTNESS register results in 100% dimming duty cycle. With this
> setting, users can just configure the color mixing register by channel to achieve the target dimming effect in a
> single-color LED application.

So this means that they provide a hint in order to drive a single LED application, just as I am doing in my specific application, but this is not directly supported by the driver. In my custom board we use three LP5024 devices. Two have RGB LEDs and a third one has just amber LEDs because the application needs that color separately.

Best,

Ivan

simonguinot reviewed (changes requested):
@ivanwagner,

Thanks for this PR. I am not against the idea of adding to the lp50xx LED driver the ability to handle single and multi channels LEDs. But I think this will require more work than the proposed patch.

Please find my review below.

simonguinot commented:
> So this means that the use case if far from being exotic and they provide a hint in order to drive a single LED application, just as I am doing in my specific application, but this is not directly supported by the Zephyr driver. In my custom board we use three LP5024 devices. Two have RGB LEDs and a third one has just amber LEDs because the application needs that color separately.
> 
> Best,
> 
> Ivan

OK. Let's see if we can make it work.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -167,14 +174,17 @@ static int lp50xx_set_color(const struct device *dev, uint32_t led,
 	}
 
+	/* Limit to RGB combination (three LEDs) */
+	if (num_colors > 3) {
```
And this should also go in a separate patch since it fixes an existing bug.

ivanwagner replied:
Is the bug you're talking about already opened? I'd be glad to provide a patch for this one in case.

simonguinot replied:
When looking at your PR I noticed that the current code the number of colors per LED found in DT is not checked during the initialization. And the code in `lp50xx_set_color` is:
```
        if (num_colors != led_info->num_colors) {
                ...
                return -EINVAL;
        }
        ...
        buf[1] = color[0];
        buf[2] = color[1];
        buf[3] = color[2];
```
The code assumes that  `num_colors` and `led_info->num_colors` are equal 3. But since it is not checked, it is not necessarily the case. So it is a bug... :)

Let it to me, I'll take care of it.

ivanwagner commented on drivers/led/lp50xx.c:
```diff
@@ -167,14 +174,17 @@ static int lp50xx_set_color(const struct device *dev, uint32_t led,
+
 	buf[0] = LP50XX_OUT0_COLOR(config->num_modules);
-	buf[0] += LP50XX_COLORS_PER_LED * led_info->index;
+	buf[0] += led_info->num_colors * led_info->index;
```
Yeah I see what you mean. As a min rule should we check that and RGB mapping is at an offset equals to a multiple of three? I think it should be the case otherwise it cannot be handled as expected. Or should this be the responsibility of the designer?

simonguinot replied:
I mean that you must state clearly what DT configurations will be supported by the driver. And during the runtime initialization the driver must check the passed DT configuration is correct.

It is the responsibility of the board designer to pass a valid configuration. But it is the responsibility of the driver to "validate" the configuration before using it. This will limit bugs and security issues. Knowing that a configuration is valid also allows to remove some checks from the API functions.

ivanwagner commented on drivers/led/lp50xx.c:
```diff
@@ -266,17 +276,26 @@ static int lp50xx_enable(const struct device *dev, bool enable)
+		num_leds += info->num_colors;
+	}
+
+	if (num_leds > config->max_leds) {
```
Where should these be documented and example provided?

simonguinot replied:
A good place would be in the DT binding. Have a look at `dts/bindings/led/issi,is31fl3194.yaml` for an example.

simonguinot replied:
And for examples, some alternatives LED nodes in a DT overlay of the `lp50xx` sample would be fine.

ivanwagner commented on include/zephyr/drivers/led/lp50xx.h:
```diff
@@ -11,12 +12,12 @@
-#define LP5024_MAX_LEDS		8
-#define LP5030_MAX_LEDS		10
-#define LP5036_MAX_LEDS		12
+#define LP5009_MAX_LEDS		9
```
I agree with you. We can change this.

ivanwagner committed:
drivers: led: lp50xx: add support for single color LEDs

Add support for single color LEDs and DT node using both on the
same controller. The index property refers to the controller
output channel for the single LED and refers to the red channel
for a RGB LED.

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

ivanwagner commented:
Hi all,

I re-propose this patch that should solve the issues mentioned by @simonguinot thus adding support for both single channel and RGB LEDs. The index property in the device tree changes though. As stated in the base LED binding documentation, it should serve as a value that maps the logical LED to the physical controller. Before it was being used as an incremental index not referring to something particular. Now it refers to the output channel number where the single channel LED is connected and regarding the RGB is refers to the connection of the red channel.

ivanwagner commented:
Can anyone take a look at this?

pdgendt commented on dts/bindings/led/ti,lp50xx.yaml:
```diff
@@ -1,14 +1,113 @@
     description: |
       GPIO to use to enable/disable the controller.
-  max-curr-opt:
+  max_curr_opt:
```
This change does not seem correct.

simonguinot reviewed (changes requested):
Hi @ivanwagner,

I deeply apologize for the delay.

I think there are many issues with this patch. Please see my review below.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -276,18 +283,19 @@ static int lp50xx_init(const struct device *dev)
 
-		if (led_info->num_colors > LP50XX_COLORS_PER_LED) {
+		used_chans += led_info->num_colors;
+
```
This path assumes that the indexes (`index` property) found in DT are coherent. It should be checked.

ivanwagner replied:
Let me see if I get it right. You mean to check if num_colors and the current index is valid on the current LED module? Do you mean to check the whole chain from index 0 to the highest configured value? Would you also check if the red LED (in a RGB module) matches the index value?

ivanwagner replied:
Ok I added a check that verifies that and RGB LED is placed on a multiple of three index as stated in the datasheet.

ivanwagner commented on drivers/led/lp50xx.c:
```diff
@@ -132,7 +132,13 @@ static int lp50xx_set_brightness(const struct device *dev,
+		buf[0] += led_info->index / led_info->num_colors;
+	} else {
+		buf[0] = LP50XX_OUT0_COLOR(config->num_modules);
+		buf[0] += led_info->index;
```
The intention was to add support for single line LEDs as stated in the PR title. I can also accommodate dual-channels LEDs as well tough.

simonguinot replied:
My point was rather than DT should be better checked to prevent unsupported configuration to be passed.

ivanwagner commented on dts/bindings/led/ti,lp50xx.yaml:
```diff
@@ -1,14 +1,113 @@
+  };
+
+  In the above example you can see how the index property is being used. The index value
+  refers to the controller output channel where the red component is being connected.
```
Yes it does break compatibility and I'd adjust other DTS across the sources that use this driver.

simonguinot replied:
You'll also have to mention this API change somewhere (maybe in `migration-guide` and `release-note`).

ivanwagner replied:
That's right

ivanwagner commented on dts/bindings/led/ti,lp50xx.yaml:
```diff
@@ -1,14 +1,113 @@
+      index = <1>;
+      color-mapping = <LED_COLOR_ID_AMBER>;
+    };
+  };
```
If the PR is accepted than your statement is not relevant anymore. Correct?

simonguinot replied:
No it still stands. But my comment is not relevant anyway :) Since DTS and driver are bound at build time, backward compatibility for DT is not needed.

ivanwagner commented on drivers/led/lp50xx.c:
```diff
@@ -351,7 +359,7 @@ static int lp50xx_pm_action(const struct device *dev,
 }
```
At page 13 from the LP5024 datasheet:

> Keeping FFh (default value) in the LED0_BRIGHTNESS register results in 100% dimming duty cycle. With this
> setting, users can just configure the color mixing register by channel to achieve the target dimming effect in a
> single-color LED application.

Am I missing something?

simonguinot replied:
No I don't think so. It looks good. Maybe just check the default value at reset of the brightness registers for the other controller models supported by this driver... Just in case :)

ivanwagner replied:
The lp50xx_reset function (called from lp50xx_init) writes to the RESET register (address 27h on the lp5024) which resets all registers to their default value including the brightness register (to FFh).

ivanwagner commented on drivers/led/lp50xx.c:
```diff
@@ -276,18 +283,19 @@ static int lp50xx_init(const struct device *dev)
 	for (led = 0; led < config->num_leds; led++) {
 		const struct led_info *led_info =
 			lp50xx_led_to_info(config, led);
 
```
In order to validate the count I need to accumulate the configured LEDs (they could both be single or RGB LEDs).

ivanwagner commented:
Anyone willing to take a look a this?

simonguinot commented on dts/bindings/led/ti,lp50xx.yaml:
```diff
@@ -1,6 +1,105 @@
+
+  Note that the RGB is connected to output 3 as suggested by the data-sheet that states that
+  the RGB's red channel should be placed on a multiple of three output (e.g. 0, 3, 6, etc).
+  So in this case index (output channel) 2 is unused.
```
It is a bit more than a suggestion but a requirement. In order to enable the use of the brightness register, the RGB LED (or any other three-channel LED) MUST be connected to an RGB LED module (i.e. three contiguous output pins and the number of the first is a multiple of three).

simonguinot commented on dts/bindings/led/ti,lp50xx.yaml:
```diff
@@ -12,7 +111,7 @@ properties:
     description: |
       If enabled the maximum current output is set to 35 mA (25.5 mA else).
-  log_scale_en:
+  log-scale-en:
```
This would be better in a separate patch.

simonguinot reviewed (changes requested):
Hi @ivanwagner,

I apologize for the delay and thank you for your continued efforts.

I must confess I am not comfortable with this PR. I agree that this driver should support single, dual and three channels LED. But the mix of channel and LED number confuses me. It looks like a hack and I don't like it. I hope that other people will review as well. This would help me to form an opinion.

Please see my comments below.

simonguinot commented:
> > Hi @ivanwagner,
> > I apologize for the delay and thank you for your continued efforts.
> > I must confess I am not comfortable with this PR. I agree that this driver should support single, dual and three channels LED. But the mix of channel and LED number confuses me. It looks like a hack and I don't like it. I hope that other people will review as well. This would help me to form an opinion.
> > Please see my comments below.
> 
> Hi Simon,
> 
> thank you very much for looking into this. Regarding the LED num vs channels, the LED num refers to the index in the DT array listing the LEDs and when we refer to channels, we are talking about the IC output. So they are two different things.
> 
> Now as I said, I am willing to work on adding this support because I need it from your side a proposal on how you'd be willing to accept a PR so that I avoid making modifications based on your feedback and once it's done I get a further comments regarding needed adjustments and in the end a general skepticism about the overall solution.
> 
> Can we make this happen? Thanks for your effort, I do appreciate it.

Hi @ivanwagner. Sorry for the delay. I'll have another look this week.

simonguinot commented:
> > Hi @ivanwagner,
> > I apologize for the delay and thank you for your continued efforts.
> > I must confess I am not comfortable with this PR. I agree that this driver should support single, dual and three channels LED. But the mix of channel and LED number confuses me. It looks like a hack and I don't like it. I hope that other people will review as well. This would help me to form an opinion.
> > Please see my comments below.
> 
> Hi Simon,
> 
> thank you very much for looking into this. Regarding the LED num vs channels, the LED num refers to the index in the DT array listing the LEDs and when we refer to channels, we are talking about the IC output. So they are two different things.
> 
> Now as I said, I am willing to work on adding this support because I need it from your side a proposal on how you'd be willing to accept a PR so that I avoid making modifications based on your feedback and once it's done I get a further comments regarding needed adjustments and in the end a general skepticism about the overall solution.
> 
> Can we make this happen? Thanks for your effort, I do appreciate it.

I read again this PR and I still stand by my original feedback: the way the index DT property refers to a channel instead of an LED a hack. On boards with RGB LEDs the indexes are now: 0, 3, 6, etc. It is not good.

Unfortunately I can't figure out a better proposition right now... The first step is designing an accurate description of the LED configuration in DT, but I have nothing to propose...

That being said, the `lp50xx` driver already supports single-color LED applications via the `->write_channel()` API. It is exactly the purpose of this API: give LED channel control to application. So maybe it is good enough ?
